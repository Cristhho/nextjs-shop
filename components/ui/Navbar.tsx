import { useContext, useState } from 'react';
import NextLink from 'next/link';
import { useRouter } from 'next/router';
import { AppBar, Badge, Box, Button, IconButton, Input, InputAdornment, Link, Toolbar, Typography } from '@mui/material';
import { ClearOutlined, SearchOutlined, ShoppingCartOutlined } from '@mui/icons-material';

import { UiContext } from '../../context';

const Navbar = () => {
  const [searchTerm, setSearchTerm] = useState<string>('');
  const [isSearchVisible, setIsSearchVisible] = useState<boolean>(false);
  const { toggleSideMenu } = useContext(UiContext);
  const { asPath, push } = useRouter();

  const onSearchHandler = () => {
    if (searchTerm.trim().length === 0) return;

    push(`/search/${searchTerm}`);
  }

  return (
    <AppBar>
      <Toolbar>
        <NextLink href='/' passHref>
          <Link display='flex' alignItems='center'>
            <Typography variant='h6'>Teslo | </Typography>
            <Typography sx={{ ml: 0.5 }}>Shop</Typography>
          </Link>
        </NextLink>
        <Box flex={1} />
        <Box sx={{
          display: isSearchVisible ? 'none' : {
            xs: 'none',
            sm: 'block'
          }
        }}>
          <NextLink href='/category/men' passHref>
            <Link>
              <Button color={`${asPath === '/category/men' ? 'primary' : 'info'}`}>Hombres</Button>
            </Link>
          </NextLink>
          <NextLink href='/category/women' passHref>
            <Link>
              <Button color={`${asPath === '/category/women' ? 'primary' : 'info'}`}>Mujeres</Button>
            </Link>
          </NextLink>
          <NextLink href='/category/kid' passHref>
            <Link>
              <Button color={`${asPath === '/category/kid' ? 'primary' : 'info'}`}>Niños</Button>
            </Link>
          </NextLink>
        </Box>
        <Box flex={1} />
        {
          isSearchVisible
          ?
            <Input
              sx={{
                display: {
                  xs: 'none',
                  sm: 'flex'
                }
              }}
              className='fadeIn'
              autoFocus
              value={searchTerm}
              type='text'
              placeholder="Buscar..."
              endAdornment={
                <InputAdornment position="end">
                  <IconButton
                    aria-label='toggle password visibility'
                    onClick={() => setIsSearchVisible(false)}
                  >
                  <ClearOutlined />
                  </IconButton>
                </InputAdornment>
              }
              onChange={(e) => setSearchTerm(e.target.value)}
              onKeyPress={(e) => e.key === 'Enter' ? onSearchHandler() : null}
            />
          :
            <IconButton
              className='fadeIn'
              sx={{
                display: {
                  xs: 'none',
                  sm: 'flex'
                }
              }}
              onClick={() => setIsSearchVisible(true)}
            >
              <SearchOutlined />
            </IconButton>
        }

        <IconButton
          sx={{ display: { xs: 'flex', sm: 'none' } }}
          onClick={toggleSideMenu}
        >
          <SearchOutlined />
        </IconButton>
        <NextLink href='/cart' passHref>
          <Link>
            <IconButton>
              <Badge badgeContent={2} color='secondary'>
                <ShoppingCartOutlined />
              </Badge>
            </IconButton>
          </Link>
        </NextLink>
        <Button onClick={() => toggleSideMenu()}>
          Menú
        </Button>
      </Toolbar>
    </AppBar>
  )
}

export default Navbar;