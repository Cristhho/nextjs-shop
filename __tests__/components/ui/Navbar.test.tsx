import { beforeEach, describe, it } from '@jest/globals';
import { render, screen, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event'

import Navbar from '@/components/ui/Navbar';

describe('<Navbar />', () => {
  it('renders without changes', () => {
    const { asFragment } = render(<Navbar />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('when Navbar component is mounted', () => {
    beforeEach(() => {
      render(<Navbar />);
    });

    it('shows the home link', () => {
    const homeLink = screen.getByRole('link', {
      name: /teslo \| shop/i
      });
  
      expect(homeLink).toBeInTheDocument();
    });
  
    it('shows the menu links', () => {
      const linksContainer = screen.getByRole('navigation');
      expect(linksContainer).toBeInTheDocument();
  
      const links = within(linksContainer).getAllByRole('link');
      expect(links).toHaveLength(3);

      const [men, women, children] = links;
      expect(men).toHaveTextContent(/Hombres/i);
      expect(women).toHaveTextContent(/Mujeres/i);
      expect(children).toHaveTextContent(/NiÃ±os/i);
    });

    it('must be a Search Button', () => {
      const search = screen.getByLabelText('Search');
      expect(search).toBeInTheDocument();
    })
  });

  describe('when the user clicks the search button', () => {
    it('the navbar links shouldn\'t be visible and the search input appears', async () => {
      render(<Navbar />);

      expect(screen.queryByRole('navigation')).toBeInTheDocument();

      userEvent.click(screen.getByLabelText('Search'));

      await waitFor(() =>
        expect(screen.queryByRole('navigation')).toBeNull()
      );

      expect(screen.getByPlaceholderText(/Buscar/i)).toBeInTheDocument();
    });
  });

});
