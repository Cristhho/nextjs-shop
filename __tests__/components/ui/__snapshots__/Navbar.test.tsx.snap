// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> renders without changes 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-32up22-MuiTypography-root-MuiLink-root"
        href="/"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
        >
          Teslo | 
        </h6>
        <p
          class="MuiTypography-root MuiTypography-body1 css-9xgob9-MuiTypography-root"
        >
          Shop
        </p>
      </a>
      <div
        class="MuiBox-root css-1rr4qq7"
      />
      <div
        class="MuiBox-root css-130f8nx"
        role="navigation"
      >
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
          href="/category/men"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1hny3k0-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Hombres
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
          href="/category/women"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1hny3k0-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Mujeres
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
          href="/category/kid"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textInfo MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1hny3k0-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Niños
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
      <div
        class="MuiBox-root css-1rr4qq7"
      />
      <button
        aria-label="Search"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium fadeIn css-16987xc-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SearchOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-112eo5r-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="SearchOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1ps4owl-MuiTypography-root-MuiLink-root"
        href="/cart"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiBadge-root BaseBadge-root css-1c32n2y-MuiBadge-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ShoppingCartOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.55 13c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.37-.66-.11-1.48-.87-1.48H5.21l-.94-2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2h7.45zM6.16 6h12.15l-2.76 5H8.53L6.16 6zM7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiBadge-badge MuiBadge-standard MuiBadge-invisible MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorSecondary BaseBadge-badge css-jcn4dz-MuiBadge-badge"
            />
          </span>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </a>
      <button
        class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Menú
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</DocumentFragment>
`;
